sklearn/gp_search.py

```diff
@@ -0,0 +1,421 @@
+from .base import is_classifier, clone
+from .grid_search import BaseSearchCV
+
+
+def _sample_candidates(n_candidates, param_bounds, param_is_int):
+    n_parameters = param_is_int.shape[0]
+    candidates = []
+
+    for k in range(n_parameters):
+        if param_is_int[k]:
+            k_sample = np.asarray(
+                np.random.rand(n_candidates)
+                * np.float(param_bounds[k][1]-param_bounds[k][0])
+                + param_bounds[k][0],
+                dtype=np.int32)
+        else:
+            k_sample = np.asarray(
+                np.random.rand(n_candidates)
+                * np.float(param_bounds[k][1]-param_bounds[k][0])
+                + param_bounds[k][0])
+        candidates.append(k_sample)
+
+    candidates = np.asarray(candidates).T
+    return unique_rows(candidates)
+
+
+def compute_expected_improvement(predictions, sigma, y_best):
+    ei_array = np.zeros(predictions.shape[0])
+    for i in range(ei_array.shape[0]):
+        z = (y_best - predictions[i]) / sigma[i]
+        ei_array[i] = sigma[i] * (z * norm.cdf(z) + norm.pdf(z))
+
+    return ei_array
+
+
+def unique_rows(ar):
+    # keep only unique values in the ndarray ar
+    # http://stackoverflow.com/questions/16970982/find-unique-rows-in-numpy-array
+
+    b = np.ascontiguousarray(ar).view(
+        np.dtype((np.void, ar.dtype.itemsize * ar.shape[1]))
+        )
+    _, idx = np.unique(b, return_index=True)
+    idx = np.sort(idx)
+
+    return ar[idx]
+
+
+def is_in_ndarray(item, a):
+    # look for element item in ndarray a
+    # returns True if item is in a, and its index
+
+    k = 0
+    idx_val = np.asarray(range(a.shape[0]))
+    idxk = range(a.shape[0])
+    while(k < a.shape[1]):
+        idxk = (a[idxk, k] == item[k])
+        if np.sum(idxk > 0):
+            k += 1
+            idx_val = idx_val[idxk]
+            idxk = list(idx_val)
+        else:
+            return False, 0
+
+    return True, idx_val[0]
+
+
+class GPSearchCV(BaseSearchCV):
+    """
+    Parameters
+    ----------
+
+    estimator : 1) sklearn estimator or 2) callable
+        1 : object type that implements the "fit" and "predict" methods,
+        as a classifier or a pipeline
+        2 : a function that computes the output given a dictionnary of
+        parameters. The returned value should be a list of one or more
+        floats if score_format == 'cv', and a float if score_format ==
+        'avg'
+
+    parameters : dict, parameter space on which to optimize the estimator
+        The keys of the dictionnary should be the names of the parameters,
+        and the values should be lists of length 2; the first element being
+        the type of the parameter ('int', 'float' or 'cat' [for categorical]),
+        and the second element being a list of either the bounds between which
+        to search (for 'int' and 'float') or the values the parameter can take
+        (for 'cat')
+        Example : parameters = {'kernel' :  ['cat', ['rbf', 'poly']],
```

Comment:
Using a namedtuple instead of a list of two entries would make the code more readable as the first and second element could be accessed by more meaningful names - like "type" and "values" rather than using expressions like "parameters[k][0] == 'float'"